### PR TITLE
Replace deprecated `wcs_renewal_order_meta` filter usage with `wc_subscriptions_renewal_order_data`

### LINKED ISSUE #640: Deprecated: Hook wcs_renewal_order_meta is <strong>deprecated</strong> since version wcs-core 2.5.0! Use wc_subscriptions_renewal_order_data instead.
Heya. This has been spamming logs now for something like a couple of years. Could we get this refactored to the suggested meta key name? Tyvm!

Today's master

https://github.com/godaddy-wordpress/wc-plugin-framework/blob/346a347196b0d6df775f293d561a73f1ae096cd8/woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php#L101

### PR DESCRIPTION
**Pull request by unfulvio-godaddy**
## Closes #640 

The filters should be compatible. We can probably roll this out contextually to #641  which this PR bases itseful upon.

I have found filter declared in Subscriptions WC_Subscriptions_Data_Copier:125-144 (see class-wc-subscriptions-data-copier.php) -- declared as follows

```php
	    /**
		 * Filters the data to be copied from one object to another.
		 *
		 * This filter name contains a dynamic part, $this->copy_type. The full set of hooks include:
		 *     - wc_subscriptions_subscription_data
		 *     - wc_subscriptions_parent_data
		 *     - wc_subscriptions_renewal_order_data
		 *     - wc_subscriptions_resubscribe_order_data
		 *
		 * @since subscriptions-core 2.5.0
		 *
		 * @param array    $data {
		 *     The data to be copied to the "to" object. Each value is keyed by the meta key. Example format [ '_meta_key' => 'meta_value' ].
		 *
		 *     @type mixed $meta_value The meta value to be copied.
		 * }
		 * @param WC_Order $from_object The object to copy data from.
		 * @param WC_Order $to_object   The object to copy data to.
		 */
		$data = apply_filters( "wc_subscriptions_{$this->copy_type}_data", $data, $this->to_object, $this->from_object );
```

### CODE REVIEW

File: woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php [hunk 1]
```diff
@@ -98,7 +98,11 @@ public function add_support() {
 		add_filter( 'woocommerce_my_subscriptions_payment_method', array( $this, 'maybe_render_payment_method' ), 10, 3 );
 
 		// don't copy over order-specific meta to the WC_Subscription object during renewal processing
-		add_filter( 'wcs_renewal_order_meta', array( $this, 'do_not_copy_order_meta' ) );
+		if ( SV_WC_Plugin_Compatibility::is_wc_subscriptions_version_gte( '2.5' ) ) {
+			add_filter( 'wc_subscriptions_renewal_order_data', [ $this, 'do_not_copy_order_meta' ] );
+		} else {
+			add_filter( 'wcs_renewal_order_meta', [ $this, 'do_not_copy_order_meta' ] );
```
Comment (lkraav) [inline woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:104]: Thanks for making this move. I also tried just renaming callback, but this triggered a bunch of undefined meta key PHP warnings. API for these callbacks maybe different, is this covered by tests?

Comment (unfulvio-godaddy) [reply woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:104]: hey @lkraav thanks for checking this out!

I was under the impression the filter had the same signature as declared in the line I pointed it out... 

May I ask which plugins are you using that rely on this FW filter?

Comment (unfulvio-godaddy) [reply woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:104]: maybe the issue is in the way the $order_meta array is shaped https://github.com/godaddy-wordpress/wc-plugin-framework/blob/1f9d73cb87732aa35172125c0e3130a22d29a871/woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php#L448 and that meta_key no longer exists?

Comment (unfulvio-godaddy) [reply woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:104]: I think  https://github.com/godaddy-wordpress/wc-plugin-framework/pull/642/commits/b37337593b1c5e57eaa6ba731e8352f506c66e22 might do the trick -- according to the updated filter description:

>  The data to be copied to the "to" object. Each value is keyed by the meta key. Example format [ '_meta_key' => 'meta_value' ]

needs testing with some concrete object, but the commit handles each possible array shape

Comment (lkraav) [reply woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:104]: > May I ask which plugins are you using that rely on this FW filter?

PayPal Gateway Powered by Braintree

https://plugins.trac.wordpress.org/browser/woocommerce-gateway-paypal-powered-by-braintree/tags/3.0.8/vendor/skyverge/wc-plugin-framework/woocommerce

File: woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php [hunk 2]
```diff
@@ -425,23 +429,25 @@ public function get_order( $order ) {
...
 	public function do_not_copy_order_meta( $order_meta ) {
 
 		$meta_keys = $this->get_order_specific_meta_keys();
 
-		foreach ( $order_meta as $index => $meta ) {
+		foreach ( (array) $order_meta as $index => $meta ) {
 
-			if ( in_array( $meta['meta_key'], $meta_keys ) ) {
+			// this accounts for different versions of the Subscriptions filter running before and after WooCommerce Subscriptions 2.5
+			if ( in_array( $index, $meta_keys ) || ( isset( $meta['meta_key'] ) && in_array( $meta['meta_key'], $meta_keys ) ) || ( isset( $meta['_meta_key'] ) && in_array( $meta['_meta_key'], $meta_keys ) ) ) {
```
Comment (lkraav) [inline woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:450]: This became quite a complicated (to read) conditional :thinking: 

* I wonder if for future generations, there might be safety in writing each piece on a separate line? Not sure if any linters would complain.

```suggestion
			if (
			        in_array( $index, $meta_keys )
			        || ( isset( $meta['meta_key'] ) && in_array( $meta['meta_key'], $meta_keys ) )
			        || ( isset( $meta['_meta_key'] ) && in_array( $meta['_meta_key'], $meta_keys ) )
			) {
```

* Having to check both `meta_key` and `_meta_key` looks quite unfortunate, are they really both in play here? I read through `WC_Subscriptions_Data_Copier::copy_data()` and `WC_Subscriptions_Data_Copier::apply_deprecated_filter()` - I could only see literal `meta_key` in code. Maybe underscore version in comments is just an example for an actual meta key, like `_paid_date` etc?

Comment (unfulvio-godaddy) [reply woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php]: https://github.com/godaddy-wordpress/wc-plugin-framework/pull/642/commits/b1418ce909946173f9596412186c68a50e2aed9d

File: woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php (hunk 1 above)
Comment (unfulvio-godaddy) [reply woocommerce/payment-gateway/integrations/class-sv-wc-payment-gateway-integration-subscriptions.php:104]: ah I see, that's maintained by A8C now -- the WC team will have to update the FW in that one after we push this change

**Review by itambek-godaddy — APPROVED** [review]
LGTM

**Pull request merged**